skimage/restoration/_denoise.py

```diff
@@ -355,3 +355,93 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
         raise ValueError('only 2-d and 3-d images may be denoised with this '
                          'function')
     return out
+
+
+def denoise_wavelet(im, threshold=None, wavelet='db1', noise_stdev=0.13):
+    """Performs wavelet denoising on an image.
+
+    Parameters
+    ----------
+    im : ndarray (2d or 3d) of ints, uints or floats
+        Input data to be denoised. `im` can be of any numeric type,
+        but it is cast into an ndarray of floats for the computation
+        of the denoised image.
+    threshold : float, optional
+        The thresholding value. All wavelet coefficients less than this value
+        are set to 0. By default, the threshold is computed adaptively as
+        described in [1]_.
+    wavelet : string, optional
+        The type of wavelet to perform. Can be any of the options
+        [pywt.wavelets]_ outputs. For example, this may be any of ``{db1, db2,
+        db3, db4}``.
+    noise_stdev : float, optional
+        The noise standard deviation used when computing the threshold
+        adaptively as described in [1]_.
+
+    Returns
+    -------
+    out : ndarray
+        Denoised image.
+
+    Notes
+    -----
+    As with the Fourier transform, there is an analogue to frequency in the
+    wavelet domain. Correspondingly, many pixel values of an image are 0 after
+    taking the wavelet transform.
+
+    By wavelet denoising, we are enforcing that
+    many of the wavelet coefficients are 0 while keeping the error small. When
+    we use soft thresholding, our estimate is
+
+    .. math:: \widehat{x} = \arg \min_x ||z - x||_2^2 + \lambda ||x||_1
+
+    where :math:`z` is the input image wavelet coefficients and :math:`\lambda`
+    is the threshold.
+
+    This function performs wavelet denoising on each color plane separately.
+
+    References
+    ----------
+    .. [wiki] https://en.wikipedia.org/wiki/Wavelet#Wavelet_denoising.
+    .. [pywt.wavedec2] http://pywavelets.readthedocs.org/en/latest/ref/2d-dwt-and-idwt.html#pywt.wavedec2
+    .. [pywt.waverec2] http://pywavelets.readthedocs.org/en/latest/ref/2d-dwt-and-idwt.html#pywt.waverec2
+    .. [pywt.families] http://pywavelets.readthedocs.org/en/latest/ref/wavelets.html#wavelet-families
+    .. [pywt] http://pywavelets.readthedocs.org/
+    .. [1] Chang, S. Grace, Bin Yu, and Martin Vetterli. "Adaptive wavelet
+           thresholding for image denoising and compression." Image Processing,
+           IEEE Transactions on 9.9 (2000): 1532-1546.
+
+    Examples
+    --------
+    >>> from skimage import color, data
+    >>> img = data.astronaut() * 1.0 / 255
+    >>> img = color.rgb2gray(img)
+    >>> img += 0.5 * img.std() * np.random.randn(*img.shape)
+    >>> img = np.clip(img, 0, 1)
+    >>> denoised_img = denoise_wavelet(img)
+    """
+    def _denoise_wavelet(img, wavelet, threshold):
```

Comment:
I believe you should move function definition outside so to avoid overhead in batch processing.